lib/tic_tac_toe.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+class TicTacToe
+    # player tokens
+    X_TOKEN = :X
+    O_TOKEN = :O
+    
+    attr_writer :board
+    attr_reader :player1
+    attr_reader :player2
+    
+    def initialize(player1, player2)
+        @board = %w[1 2 3 4 5 6 7 8 9]
+        @player1 = player1
+        @player2 = player2
+    end
+
+    private
+    
+    # This method is used to validate if an position is already taken by a player
+    # @param index
+    # @return boolean
+    def validate_index(index)
+        if @board[index] == X_TOKEN || @board[index] == O_TOKEN
+            false
+        else
+            true
+        end
+    end
+
+    # this method is used to check if the given element is a valid board position
+    # @param position
+    # @return boolean
+    def board_position(position)
+        if position.to_i >= 0 && position.to_i < 9
+            true
+        else
+            false
+        end
+    end
+    
+    def rows
+        # board rows
+        row1 = [@board[0], @board[1], @board[2]]
+        row2 = [@board[3], @board[4], @board[5]]
+        row3 = [@board[6], @board[7], @board[8]]
+        [row1, row2, row3]
+    end
+    
+    def cols
+        # board columns
+        col1 = [@board[0], @board[3], @board[6]]
+        col2 = [@board[1], @board[4], @board[7]]
+        col3 = [@board[2], @board[5], @board[8]]
+        [col1, col2, col3]
+    end
+    
+    def diagonals
+        # board diagonals
+        diag1 = [@board[0], @board[4], @board[8]]
+        diag2 = [@board[2], @board[4], @board[6]]
+        [diag1, diag2]
+    end    
+
+    # this method is used to evaluate if a winner combination is found
+    # @return boolean
+    def combinations
+        winner = false
+        data = {}
+    
```

Comment:
Every method should do just one thing. Try to see it like this, try to define what is this method doing, if describing what the method does you have to incorporate an "and", then it means it's too much and it should be separated in two methods.


> Methods, like classes, should have a single responsibility. All of the same reasons
> apply; having just one responsibility makes them easy to change and easy to reuse. All
> the same design techniques work; ask them questions about what they do and try to
> describe their responsibilities in a single sentence. 

> 
>  Methods that have a single responsibility confer the following benefits:
> • **Expose previously hidden qualities** Refactoring a class so that all of its methods
> have a single responsibility has a clarifying effect on the class. Even if you do not
> intend to reorganize the methods into other classes today, having each of them
> serve a single purpose makes the set of things the class does more obvious.
> • **Avoid the need for comments** How many times have you seen a comment that is
> out of date? Because comments are not executable, they are merely a form of
> decaying documentation. If a bit of code inside a method needs a comment,
> extract that bit into a separate method. The new method name serves the same
> purpose as did the old comment.
> • **Encourage reuse** Small methods encourage coding behavior that is healthy for
> your application. Other programmers will reuse the methods instead of duplicating the code. They will follow the pattern you have established and create small,
> reusable methods in turn. This coding style propagates itself.
> • **Are easy to move to another class** When you get more design information and
> decide to make changes, small methods are easy to move. You can rearrange behavior
> without doing a lot of method extraction and refactoring. Small methods lower
> the barriers to improving your design.


_**ref: Practical Object-Oriented Design in Ruby: An Agile Primer by Sandi Metz**_